### PR TITLE
12.0 [IMP] beesdoo_shift_swap

### PR DESCRIPTION
**Pull request by MatteoLdx**
## Description

Improvements mostly for modules beesdoo_shift_swap and beesdoo_website_shift_swap.

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

### CODE REVIEW

**Comment by remytms** [comment]
@MatteoLdx You can rebase on top of `12.0`. :)

File: beesdoo_shift_swap/data/mail_template.xml [hunk 1]
```diff
@@ -170,5 +170,38 @@
...
+                    <p>
+                        If you wish to do so, please <a href="/my/shift/swaping/${object.exchanged_tmpl_dated_id.template_id.id}/${object.exchanged_tmpl_dated_id.date.strftime('%Y-%m-%d %H:%M:%S')}">click here</a>.
+                    </p>
+                    <br/><br/>
+                    Regards,
+                </div>
+                ]]>
+            </field>
+        </record>
+
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/data/mail_template.xml]:
```suggestion
        <!-- Contact coop if no match has been found for his/her exchange request -->
        <record id="email_template_warn_user_no_match" model="mail.template">
            <field
                name="name"
            >Contact coop if no match has been found for his/her exchange request</field>
            <field
                name="model_id"
                ref="beesdoo_shift_swap.model_beesdoo_shift_exchange_request"
            />
            <field name="partner_to">${object.worker_id.id}</field>
            <field name="auto_delete" eval="False" />
            <field
                name="subject"
            >No match has been found for your exchange request</field>
            <field name="body_html">
                <![CDATA[
                <div style="font-family: 'Lucida Grande', Ubuntu, Arial, Verdana, sans-serif; font-size: 12px; color: rgb(34, 34, 34); background-color: #FFF; ">
                    Dear ${object.worker_id.name},
                    <br/><br/>
                    <p>
                        You are receiving this email because no match has been found for your exchange request, and the shift that you want
                        to exchange is getting closer. To avoid missing it, you can now swap it with an underpopulated one.
                    </p>
                    <p>
                        If you wish to do so, please <a href="/my/shift/swaping/${object.exchanged_tmpl_dated_id.template_id.id}/${object.exchanged_tmpl_dated_id.date.strftime('%Y-%m-%d %H:%M:%S')}">click here</a>.
                    </p>
                    <br/><br/>
                    Regards,
                </div>
                ]]>
            </field>
        </record>
```
Fixing typos and language. Can you make the same changes in other templates in this file ?

Comment (victor-champonnois) [reply beesdoo_shift_swap/data/mail_template.xml]: Fixing typos and language

File: beesdoo_shift_swap/wizard/exchange_wizard.py [hunk 2]
```diff
@@ -134,13 +138,17 @@ def make_change(self):
...
-        return True
+        self = self._check()
+        if not self.asked_tmpl_dated_ids:
+            raise UserError(_("Please ask for at least one shift"))
+        for tmpl_dated in self.asked_tmpl_dated_ids:
+            for worker in tmpl_dated.template_id.worker_ids:
+                self.worker_id.send_mail_for_exchange(
+                    self.exchanged_tmpl_dated_id, tmpl_dated, worker
+                )
+        self.make_change()
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/wizard/exchange_wizard.py]: Could this method have a more precise name ? Like make_exchange ?

File: beesdoo_shift_swap/wizard/exchange_wizard.py [hunk 3]
```diff
@@ -108,9 +108,13 @@ def _check(self, group="beesdoo_shift.group_shift_management"):
             raise UserError(_("You cannot perform this operation on yourself"))
         return self.with_context(real_uid=self._uid)
 
-    @api.multi
     def make_change(self):
         self = self._check()
+        if not self.asked_tmpl_dated_ids or not self.possible_match:
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/wizard/exchange_wizard.py]: Isn't it a `and` that is needed here, rather than `or` ?

File: beesdoo_shift_swap/models/res_partner.py [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
 
 class ResPartner(models.Model):
 
     _inherit = "res.partner"
 
+    subscribed_exchange_emails = fields.Boolean(
+        string="Echange emails subscription",
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/res_partner.py]:
```suggestion
         string="Exchange emails subscription",
```

File: beesdoo_shift_swap/wizard/exchange_wizard.py (hunk 3 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/wizard/exchange_wizard.py]: You're right, I already fixed it [here](https://github.com/beescoop/Obeesdoo/pull/422/commits/709beb78af07761585248c51f8753a46520b2a8e)

File: beesdoo_website_shift_swap/views/general_templates.xml [hunk 5]
```diff
@@ -208,6 +256,34 @@
...
+            name="Message after mail subscription update"
+        >
+            <t t-call="portal.portal_layout">
+                <t t-set="no_breadcrumbs" t-value="True" />
+                <div class="container mt16">
+                    <div class="row">
+                        <div class="col text-center">
+                            <div class="alert alert-success">
+                                <div t-if="is_subscribed">
+                                    You have been successfully unsubscribed to exchange-related emails.
```
Comment (victor-champonnois) [inline beesdoo_website_shift_swap/views/general_templates.xml]:
```suggestion
                                     You have been successfully unsubscribed from exchange-related emails.
```
~~and same below~~

File: beesdoo_shift_swap/models/res_partner.py [hunk 6]
```diff
@@ -226,9 +226,7 @@ def exchange_shifts(
             if not new_shift_generated:
                 new_shift = wanted_tmpl_dated.new_shift(self)
                 if solidarity_offer:
-                    new_shift.write(
-                        {"solidarity_offer_ids": [(6, 0, solidarity_offer.ids)]}
-                    )
+                    new_shift.solidarity_offer_ids = [(6, 0, solidarity_offer.ids)]
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/res_partner.py]: What is the reason for this change?

Comment (MatteoLdx) [reply beesdoo_shift_swap/models/res_partner.py]: It didn't work with write(), don't know why

File: beesdoo_shift_swap/wizard/exchange_wizard.py [hunk 7]
```diff
@@ -54,14 +56,21 @@ def _get_template_dated(self):
     @api.onchange("worker_id")
     def _get_available_tmpl_dated(self):
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/wizard/exchange_wizard.py]: Maybe change this method name to prevent a confusion with `get_available_tmpl_dated`

File: beesdoo_shift_swap/data/mail_template.xml (hunk 1 above)
Comment (victor-champonnois) [reply beesdoo_shift_swap/data/mail_template.xml]: actually you can ignore the "Dear ...,"

File: beesdoo_shift_swap/models/planning.py [hunk 8]
```diff
@@ -13,78 +13,103 @@ def _prepare_task_day(self):
...
+        # Get all the changes and store them into a list to sort them
+        changes = []
+
+        swaps = self.env["beesdoo.shift.swap"].search(
+            [
+                ("state", "=", "validated"),
+            ]
+        )
+        for swap in swaps:
+            changes.append(swap)
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/planning.py]: Why not just doing `list(swaps)` ?

File: beesdoo_shift_swap/models/planning.py [hunk 9]
```diff
@@ -13,78 +13,103 @@ def _prepare_task_day(self):
...
+
         exchanges = self.env["beesdoo.shift.exchange"].search([])
-        solidarity_offers = self.env["beesdoo.shift.solidarity.offer"].search([])
-        solidarity_requests = self.env["beesdoo.shift.solidarity.request"].search([])
+        for exchange in exchanges:
+            changes.append(exchange)
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/planning.py]: also here just do `changes + list(exchanges)` ?

Comment (victor-champonnois) [reply beesdoo_shift_swap/models/planning.py]: I guess it's faster than using a loop?

File: beesdoo_shift_swap/data/mail_template.xml (hunk 1 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/data/mail_template.xml]: Done

File: beesdoo_shift_swap/wizard/exchange_wizard.py (hunk 2 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/wizard/exchange_wizard.py]: Done

File: beesdoo_shift_swap/models/res_partner.py (hunk 4 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/models/res_partner.py]: Done

File: beesdoo_website_shift_swap/views/general_templates.xml (hunk 5 above)
Comment (MatteoLdx) [reply beesdoo_website_shift_swap/views/general_templates.xml]: Done

File: beesdoo_shift_swap/wizard/exchange_wizard.py (hunk 7 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/wizard/exchange_wizard.py]: Done

File: beesdoo_shift_swap/models/planning.py (hunk 8 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/models/planning.py]: Done

File: beesdoo_shift_swap/models/planning.py (hunk 9 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/models/planning.py]: Done

File: beesdoo_shift_swap/models/res_partner.py [hunk 10]
```diff
@@ -4,6 +4,13 @@
 from odoo.exceptions import UserError
 
 
+def get_validate_date(element):
+    if element.__class__.__name__ == "beesdoo.shift.exchange_request":
+        return element.validate_date
+    else:
+        return element.create_date
+
+
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/res_partner.py:13]: this seems unusual, let's talk about this together :)

File: beesdoo_shift_swap/models/dated_template.py [hunk 11]
```diff
@@ -13,11 +13,20 @@ class DatedTemplate(models.Model):
...
     end_time = fields.Datetime(required=True, compute="_compute_time")
 
+    @api.depends("date")
+    def _compute_str_date(self):
+        for record in self:
+            if not record.date:
+                record.str_date = False
+            else:
+                record.str_date = record.date.strftime("%d/%m/%Y")
+
```
Comment (victor-champonnois) [inline beesdoo_shift_swap/models/dated_template.py]: This should not be done using python, but the date format must be changed in Odoo settings (see [here](https://www.odoo.com/forum/help-1/how-can-we-change-date-view-format-from-dd-mm-yyyy-to-mm-dd-yyyy-28299)). Therefore this commit is not needed.
You should talk to an analyst about this.

File: beesdoo_shift_swap/models/planning.py [hunk 12]
```diff
@@ -13,78 +13,103 @@ def _prepare_task_day(self):
...
-                    not shift["worker_id"]
-                    and swap.wanted_tmpl_dated_id.template_id.id
-                    == shift["task_template_id"]
-                    and shift["start_time"] == swap.wanted_tmpl_dated_id.date
-                ):
-                    if template["initial"] != template["exchange_modified"]:
-                        shift["worker_id"] = swap.worker_id.id
+            for rec in changes:
+                class_name = rec.__class__.__name__
+                if class_name == "beesdoo.shift.swap":
```
Comment (robinkeunen) [inline beesdoo_shift_swap/models/planning.py]: This raises many warning bells : couldn't you rather split this long function into smaller ones and specialise in inheriting classes ?

File: beesdoo_shift_swap/models/res_partner.py [hunk 13]
```diff
@@ -1,11 +1,25 @@
-from odoo import _, api, models
+from datetime import datetime
+
+from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
 
+def get_validate_date(element):
```
Comment (robinkeunen) [inline beesdoo_shift_swap/models/res_partner.py]: As @victor-champonnois suggests, it's probably better to have this function as a private static method.
Also, couldn't you overide this function in class `beesdoo.shift.exchange_request` to have it use `validate_date`?

**Review by robinkeunen — CHANGES_REQUESTED** [review]

File: beesdoo_website_shift_swap/views/general_templates.xml [hunk 14]
```diff
@@ -26,6 +26,53 @@
             </div>
         </template>
 
+        <!-- Mailing list button -->
+        <template
+            id="manage_emails_button"
+            name="Button to manage email subscription"
+            inherit_id="portal.portal_my_home"
+        >
+            <xpath expr="//div[hasclass('o_portal_my_home')][1]" position="inside">
```
Comment (victor-champonnois) [inline beesdoo_website_shift_swap/views/general_templates.xml:35]: We can look together maybe there is a better xpath to use here

File: beesdoo_website_shift_swap/views/general_templates.xml [hunk 15]
```diff
@@ -143,28 +216,26 @@
...
             </xpath>
         </template>
 
+        <!-- Add label for solidarity shifts -->
+        <template
+            id="shift_swap_status_label"
+            name="Shift Swap Status Label"
+            inherit_id="beesdoo_website_shift.shift_status_label"
+        >
+            <xpath expr="//span[last()]" position="after">
```
Comment (victor-champonnois) [inline beesdoo_website_shift_swap/views/general_templates.xml:246]: also here, span seems very generic to use in an xpath

File: beesdoo_shift_swap/models/dated_template.py (hunk 11 above)
Comment (MatteoLdx) [reply beesdoo_shift_swap/models/dated_template.py]: Done

**Review by robinkeunen — APPROVED** [review]
Only reviewed yesterday's and today's commits. MUCH MORE SATISFYING.

**Comment by robinkeunen** [comment]
@remytms @MatteoLdx attention de bien squasher les fixups avant de merger 🧹
